base/applications/rapps/asyncinet.cpp

```diff
@@ -0,0 +1,258 @@
+            }
+            else // asynchronous InternetReadFile complete
+            {
+                AsyncInetPerformCallback(AsyncInet, ASYNCINET_DATA, (WPARAM)(AsyncInet->ReadBuffer), (LPARAM)(AsyncInet->BytesRead));
+            }
+
+            while (1)
+            {
+                BOOL bRet = InternetReadFile(AsyncInet->hInetFile,
+                    AsyncInet->ReadBuffer,
+                    _countof(AsyncInet->ReadBuffer),
+                    &(AsyncInet->BytesRead));
+                if (bRet)
+                {
+                    if (AsyncInet->BytesRead == 0)
+                    {
+                        // all read.
+                        AsyncInetPerformCallback(AsyncInet, ASYNCINET_COMPLETE, 0, 0);
+
+                        // clean up, close handles.
+
+                        // AsyncInet may got freed while handling INTERNET_STATUS_HANDLE_CLOSING.
+                        // so store the handle first, then close them
+                        HINTERNET hInetFile = AsyncInet->hInetFile;
+                        HINTERNET hInternet = AsyncInet->hInternet;
+                        if (hInetFile)
+                        {
+                            InternetCloseHandle(hInetFile);
+                        }
+                        if (hInternet)
+                        {
+                            InternetCloseHandle(hInternet);
+                        }
+                        break;
+                    }
+                    else
+                    {
+                        // read completed immediately.
+                        AsyncInetPerformCallback(AsyncInet, ASYNCINET_DATA, (WPARAM)(AsyncInet->ReadBuffer), (LPARAM)(AsyncInet->BytesRead));
+                    }
+                }
+                else
+                {
+                    DWORD dwError;
+                    if ((dwError = GetLastError()) == ERROR_IO_PENDING)
+                    {
+                        // performing asynchronous IO, everything OK.
+                        break;
+                    }
+                    else
+                    {
+                        //something went wrong
+                        if (dwError == ERROR_INVALID_HANDLE)
+                        {
+                            // That's most likely not an error. handle got closed when canceling may lead to this
+                            break;
+                        }
+                        if (AsyncInet->Callback)
+                        {
+                            AsyncInet->Callback(AsyncInet, ASYNCINET_ERROR, 0, (LPARAM)dwError, AsyncInet->Extension);
+                        }
+                        break;
+                    }
+                }
+            }
+        }
+        break;
+        case ERROR_INVALID_HANDLE:
+            if (AsyncInet->bIsCancelled)
+            {
+                // this is most likely not an error. when canceling, using closed/invalid handle may lead to this
```

Comment:
I would say this is up to the class calling the asyncinet?
In some cases this might be true, in other cases this will be an actual error during transmission.

Reply:
yes and now I add code to verify AsyncInet->bIsCancelled. Is this enough?